feat: Improve error handling

Richterrettich opened the pull request:
The error type is now an enum with variants for different error scenarios.
To reduce boilerplate code the _thiserror_ crate was used.
Errors intended for signature handling are more generic because it is
likely that library users will implement a custom signer at some point
in time.

Fixes #14

Richterrettich commented:
@drahnr , do you want to take a look? Tell me if you think there are problems with the implementations or if something is missing.

drahnr commented on src/errors.rs:
```diff
@@ -1,56 +1,93 @@
-}
+    #[error("error verifying signature with key {key_ref}: {source}")]
+    VerificationError {
+        source: Box<dyn std::error::Error>,
```
:+1:

drahnr commented on src/rpm/builder.rs:
```diff
@@ -144,17 +140,20 @@ impl RPMBuilder {
-            .ok_or_else(|| RPMError::new(&format!("invalid destination path {}", dest)))?;
+            .ok_or_else(|| RPMError::InvalidDestinationPath {
+                path: dest.clone(),
+                desc: "does not have a parent directory",
```
We should consider phrasing those in a consistent way (capitalization, active/passive form).

drahnr commented on src/rpm/headers/header.rs:
```diff
@@ -123,54 +129,67 @@ where
             .iter()
             .find(|entry| &entry.tag == tag)
-            .ok_or_else(|| RPMError::new(&format!("unable to find Tag {}", tag)))
+            .ok_or_else(|| RPMError::TagNotFound(format!("{}", tag)))
```
```suggestion
            .ok_or_else(|| RPMError::TagNotFound(tag.to_string()))
```

drahnr replied:
I prefer the the `.to_string()` variants, since it introduces less visual clutter and does the same.

Richterrettich replied:
yeah, I completely forgot that `to_string` was implied by the implemented traits.

drahnr commented on src/rpm/headers/header.rs:
```diff
@@ -123,54 +129,67 @@ where
+            .as_str()
+            .ok_or_else(|| RPMError::UnexpectedTagDataType {
+                expected_data_type: "string",
+                actual_data_type: format!("{}", entry.data),
+                tag: format!("{}", entry.tag),
```
```suggestion
                actual_data_type: entry.data.to_string(),
                tag: entry.tag.to_string(),
```

drahnr commented on src/rpm/headers/header.rs:
```diff
@@ -123,54 +129,67 @@ where
+            .ok_or_else(|| RPMError::UnexpectedTagDataType {
+                expected_data_type: "i32",
+                actual_data_type: format!("{}", entry.data),
+                tag: format!("{}", entry.tag),
```
ff as before

drahnr commented on src/tests.rs:
```diff
@@ -261,7 +261,7 @@ fn test_region_tag() -> Result<(), Box<dyn std::error::Error>> {
     let data = match region_entry.data {
         IndexData::Bin(d) => d,
-        _ => return Err(Box::new(RPMError::new("should be bin"))),
+        _ => return Err(Box::new(RPMError::Other("should be bin".to_string()))),
```
Avoiding generic errors would be a big plus, every variant should have one root cause. A `other` is the uneeded escape hatch here. I'd rather have this be named specifically to this use-case.

drahnr commented on src/rpm/signature/pgp.rs:
```diff
@@ -46,7 +46,8 @@ impl traits::Signing<traits::algorithm::RSA> for Signer {
             ]) // must be initialized
-            .build()?;
+            .build()
+            .map_err(|e| RPMError::SignError(Box::new(RPMError::Other(e))))?;
```
Can we have a `RPMError::Signing(e: sd::error::Error)` instead?

Richterrettich replied:
I thought so too but it is not as easy since `e` in this case is just a `String`. Thats one of the only reasons I introduced  the general `Other` error.
The alternative would be to create a custom (private) error type within the `pgp` crate to wrap this error string to make it a proper error.
Do you see any other solution?

Richterrettich commented:
I've removed the generic `Other` error type and created String wrapper where necessary. Let me know if you have a better idea how to solve this properly.

drahnr commented:
> I've removed the generic `Other` error type and created String wrapper where necessary. Let me know if you have a better idea how to solve this properly.
> I did not want to create public errors that are only created for internal use. This seems fishy that's why I created explicit error types for those.

They would become public by indirection, when they are the `source` of an error that then becomes public. I don't see an issue with the fact that the are never the error variant that is returned by a function.

drahnr commented:
Other than that, LGTM :+1:

drahnr commented on src/tests.rs:
```diff
@@ -261,7 +261,11 @@ fn test_region_tag() -> Result<(), Box<dyn std::error::Error>> {
-        _ => return Err(Box::new(RPMError::new("should be bin"))),
+        _ => {
+            return Err(Box::new(TestError::StringError(
+                "should be bin".to_string(),
```
Why bother to return an error here? Can't we simply it with an `expect_matches!()` from crate `expect_matches` and use expects or `assert!(...is_ok())` / `.expect("yada")` otherwise?

drahnr commented on src/tests.rs:
```diff
@@ -272,3 +276,12 @@ fn test_region_tag() -> Result<(), Box<dyn std::error::Error>> {
+enum TestError {
+    #[error("{0}")]
+    StringError(String),
+}
```
.. makes this superfluous.

Richterrettich commented:
I've restructured the code and removed the error from the test case as you have suggested.
For the PGP package I did a similar thing. The builder is a generated construct from a crate called `rust_derive_builder` (this is why my IDE freaks out about the builder too btw.)
The `build` method always returns a String error but it is mostly just there for convenience. 
I've removed the builder and initialized the struct directly. As a result, there is no more need for an error wrapper.

Do you want to have a final look?

drahnr commented:
LGTM!

drahnr commented on src/rpm/signature/pgp.rs:
```diff
@@ -336,22 +341,21 @@ pub(crate) mod test {
-            .hashed_subpackets(vec![
-                ::pgp::packet::Subpacket::SignatureCreationTime(now),
-                ::pgp::packet::Subpacket::Issuer(signer.secret_key.key_id()),
+
```
The only downside I see here is the kids of sanitization checks `.build()` would do assuming it does some.

drahnr commented on src/rpm/signature/pgp.rs:
```diff
@@ -32,35 +34,34 @@ impl traits::Signing<traits::algorithm::RSA> for Signer {
-            ]) // must be initialized
-            .build()?;
 
+        let sig_cfg = SignatureConfig {
```
Same as further up

drahnr commented on src/tests.rs:
```diff
@@ -259,10 +259,11 @@ fn test_rpm_header() -> Result<(), Box<dyn std::error::Error>> {
-    };
+    let possible_binary = region_entry.data.as_binary();
+
+    assert!(possible_binary.is_some(), "should be binary");
```
`Option` can also use ` .expect()` iirc
